compiler/rustc_mir_build/src/builder/matches/match_pair.rs

```diff
@@ -333,3 +334,9 @@ impl<'tcx> MatchPairTree<'tcx> {
         }
     }
 }
+
+/// Sorts or-patterns to the end of the list, as required by [`FlatPat`] and
+/// its enclosed [`MatchPairTree`] nodes.
+pub(super) fn sort_match_pairs(match_pairs: &mut [MatchPairTree<'_>]) {
+    match_pairs.sort_by_key(|pair| matches!(pair.test_case, TestCase::Or { .. }));
+}
```

Comment:
While you're at it, could you also replace the `sort_by_key` call here: https://github.com/rust-lang/rust/blob/4559163ccb500affc424fb9228dae5003672ffc7/compiler/rustc_mir_build/src/builder/matches/test.rs?plain=1#L772